File: flink-java-code/src/main/java/com/atguigu/day3/RichFunctionExample.java
package com.atguigu.day3;

import org.apache.flink.api.common.functions.RichMapFunction;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

public class RichFunctionExample {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);

        DataStream<Integer> stream = env.fromElements(1,2,3);

        stream
                .map(new RichMapFunction<Integer, Integer>() {
                    @Override
                    public void open(Configuration parameters) throws Exception {
                        System.out.println("enter lifecycle");
                    }

                    @Override
                    public Integer map(Integer value) throws Exception {
                        return value + 1;
                    }

                    @Override
                    public void close() throws Exception {
                        System.out.println("exit lifecycle");
                    }
                })
                .print();

        env.execute();
    }
}

File: flink-scala-code/src/main/scala/com/atguigu/day3/RichFunctionExample.scala
package com.atguigu.day3

import org.apache.flink.api.common.functions.RichMapFunction
import org.apache.flink.configuration.Configuration
import org.apache.flink.streaming.api.scala._

object RichFunctionExample {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)

    val stream = env.fromElements(1,2,3)

    stream
      .map(new MyMapFunction)
      .print()
    env.execute()
  }

  class MyMapFunction extends RichMapFunction[Int, Int] {
    override def open(parameters: Configuration): Unit = {
      println("enter into lifecycle")
    }

    override def map(value: Int): Int = {
      value + 1
    }

    override def close(): Unit = {
      println("exit lifecycle")
    }
  }
}